moveit_ros/moveit_servo/include/moveit_servo/input_pipeline.hpp

```diff
@@ -0,0 +1,117 @@
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *******************************************************************************/
+
+/*      Title     : input_pipeline.hpp
+ *      Project   : moveit_servo
+ *      Created   : 11/06/2021
+ *      Author    : Tyler Weaver
+ */
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <moveit_servo/input_pipeline.hpp>
+#include <moveit_servo/input_resampler.hpp>
+#include <moveit_servo/input_stale_command.hpp>
+#include <moveit_servo/input_subscriber.hpp>
+#include <moveit_servo/servo_parameters.h>
+#include <rclcpp/rclcpp.hpp>
+
+namespace moveit_servo
+{
+class ResamplingInputPipeline
+{
+  // TODO (tylerjw): make so the address of these can change
+  std::unique_ptr<StaleCommandHalt> stale_command_halt_;
+  std::unique_ptr<InputResampler> input_resampler_;
+  std::unique_ptr<TimestampNow> timestamp_now_;
+  InputSubscriber input_subscriber_;
+
+public:
+  ResamplingInputPipeline(rclcpp::Node::SharedPtr node, std::shared_ptr<const moveit_servo::ServoParameters> parameters,
+                          const std::function<void()>& halt, InputVisitor* next)
+    : stale_command_halt_{ std::make_unique<StaleCommandHalt>(
+          node, rclcpp::Duration::from_seconds(parameters->incoming_command_timeout), halt, next) }
+    , input_resampler_{ std::make_unique<InputResampler>(
+          node, rclcpp::Duration::from_seconds(parameters->publish_period), stale_command_halt_.get()) }
+    , timestamp_now_{ std::make_unique<TimestampNow>(node, input_resampler_.get()) }
+    , input_subscriber_{ node, parameters->cartesian_command_in_topic, parameters->joint_command_in_topic,
+                         timestamp_now_.get() }
+  {
+  }
+};
+
+class ReactiveInputPipeline
+{
+  InputSubscriber input_subscriber_;
+
+public:
+  ReactiveInputPipeline(rclcpp::Node::SharedPtr node, std::shared_ptr<const moveit_servo::ServoParameters> parameters,
+                        InputVisitor* next)
+    : input_subscriber_{ node, parameters->cartesian_command_in_topic, parameters->joint_command_in_topic, next }
+  {
+  }
+};
+
+using InputPipeline = std::variant<ResamplingInputPipeline, ReactiveInputPipeline>;
+
+/**
+ * @brief      Builds the input pipeline
+ *
+ * @param[in]  node        The ros node
+ * @param[in]  parameters  The parameters
+ * @param[in]  halt        The halt function
+ * @param[in]  next        The next visitor connect the pipeline to
+ *
+ * @return     The input pipeline.
+ */
+InputPipeline makeInputPipeline(rclcpp::Node::SharedPtr node,
+                                std::shared_ptr<const moveit_servo::ServoParameters> parameters,
+                                const std::function<void()>& halt, InputVisitor* next)
+{
+  assert(node != nullptr);
+  assert(parameters != nullptr);
+  assert(next != nullptr);
+
+  if (parameters->low_latency_mode)
+  {
+    return ReactiveInputPipeline{ node, parameters, next };
```

Comment:
How is `halt` implemented for this pipeline?

Reply:
It is not handled.  In my experimentation at least if you tie the output in a 1:1 relationship with input servo will not continue to move the robot.  This is because it sends small steps to the robot and once the robot achieves that step it doesn't continue.  This even seems to be true if you send a twist with a velocity.